aten/src/ATen/native/quantized/cpu/BinaryOps.cpp

```diff
@@ -393,25 +393,19 @@ Tensor xnnp_add(Tensor qa, Tensor qb, double scale, int64_t zero_point) {
 template <bool ReLUFused = false>
 Tensor qadd(Tensor qa, Tensor qb, double scale, int64_t zero_point) {
   check_inputs(qa, qb);
-
-  if (at::globalContext().qEngine() == at::QEngine::QNNPACK) {
```

Comment:
Well..I would argue that this is not incorrect, but it wasn't very clear either.

The context is, we use same qengine for both xnnpack and qnnpack. The rationale was that there weren't big enough feature differences between the two which warrants introducing new qengine. However, when you use qnnpack qengine with symmetric quantization schema, you "transparently" pick xnnpack and similarly for asymmetric you would pick qnnpack, hence the check on the dtype being signed for xnnpack and unsigned for qnnpack.

Reply:
Thank you for providing the context. I erroneously assumed the engine check is incorrectly placed. After digging a bit deeper, it looks like qnnpack engine is only set for mobile ? 

https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/Context.cpp#L379-L403

Furthermore, when MKLDNN is enabled the qnnpack engine will be reset to onednn engine. So, how can we leverage xnnpack for this op when we build with `AT_MKLDNN_ENABLED()` ? Could you please advise ? Thank you.

Reply:
Yeah that's a good question. Typically the kernels are written with certain assumptions around quantization, hence the need for the qengine and kernels specific to those. 

Yes qengine==qnnpack is typically for mobile (i.e. Arm cpus) because OneDNN has arguably better x86 quantized kernels while qnnpack/xnnpack has been focusing on Arm more.

Also IIRC x86 qschemas were uint8_t i.e. [0, 255] while this is [-128, 127] quant value range. THat's why I was asking how did you quantize your operator?

Reply:
I quantized `torch.nn.quantized.FloatFunctional().add` with this QConfig
```
    qconfig = QConfig(
        activation=HistogramObserver.with_args(qscheme=torch.per_tensor_affine, dtype=torch.qint8),
        weight=MinMaxObserver.with_args(dtype=torch.qint8, qscheme=torch.per_tensor_symmetric)
    )
```
with `torch.ao.quantization.convert`.

So, back to the question, how should I enable xnnpack on Arm server CPU for this op ?

Reply:
@digantdesai Do you have further comments on this ? Thank you.

Reply:
So you are not specifying backend or qengine. Not all CPU backends can handle this qdata_type + qschema combination. See this - https://pytorch.org/docs/stable/quantization.html#note-for-native-cpu-backends


> how should I enable xnnpack on Arm server CPU for this op ?
qnnpack backend (or qengine) used with int8 dtype (as opposed to uint8) will pick xnnpack in eager mode. For PT2E quant flow you can use XNNPACKQuantizer. 

See this - https://pytorch.org/tutorials/prototype/pt2e_quant_ptq.html

Reply:
I would like to use oneDNN backend when running quantized linear op as well. So, this is actually not the right solution. Plus, it turns out that XNNPACK runs slower with a single core on the system I am testing than aten implementation. I will look for a different solution. Thank you for helping me get a better understanding 😄